pkg/protocol/engine/attestation/slotattestation/manager.go

```diff
@@ -0,0 +1,249 @@
+		if !exists {
+			return newAttestation
+		}
+
+		// Replace the attestation only if the new one is greater.
+		if newAttestation.Compare(currentValue) == 1 {
+			return newAttestation
+		}
+
+		return currentValue
+	})
+}
+
+func (m *Manager) applyToPendingAttestations(attestation *iotago.Attestation, cutoffIndex iotago.SlotIndex) {
+	if attestation.SlotCommitmentID.Index() < cutoffIndex {
+		return
+	}
+
+	updated := true
+
+	updatedAttestation := m.pendingAttestations.Get(cutoffIndex, true).Compute(attestation.IssuerID, func(currentValue *iotago.Attestation, exists bool) *iotago.Attestation {
+		if !exists {
+			return attestation
+		}
+
+		// Replace the stored attestation only if the new one is greater.
+		if attestation.Compare(currentValue) == 1 {
+			return attestation
+		}
+
+		updated = false
+
+		return currentValue
+	})
+
+	// The new attestation is smaller or equal the previousAttestation. There's no need to update pendingAttestations.
+	if !updated {
+		return
+	}
+
+	for i := cutoffIndex; i <= updatedAttestation.SlotCommitmentID.Index(); i++ {
+		m.pendingAttestations.Get(i, true).Set(attestation.IssuerID, updatedAttestation)
+	}
+}
+
+func (m *Manager) determineAttestationsFromWindow(slotIndex iotago.SlotIndex) []*iotago.Attestation {
+	slotAttestors := m.pendingAttestations.Get(slotIndex, false)
+	if slotAttestors == nil {
+		return nil
+	}
+
+	return slotAttestors.Values()
+}
+
+func (m *Manager) Commit(index iotago.SlotIndex) (newCW uint64, attestationsRoot types.Identifier, err error) {
+	m.commitmentMutex.Lock()
+	defer m.commitmentMutex.Unlock()
+
+	cutoffIndex, valid := m.computeAttestationCommitmentOffset(index)
+
+	// Remove all future attestations of slot and apply to pending attestations window.
+	futureAttestations := m.futureAttestations.Evict(index)
+	if futureAttestations != nil {
+		futureAttestations.ForEach(func(key iotago.AccountID, value *iotago.Attestation) bool {
+			m.applyToPendingAttestations(value, cutoffIndex)
+
+			return true
+		})
+	}
+
+	if !valid {
+		m.lastCommittedSlot = index
+		return 0, types.Identifier{}, nil
+	}
+
+	// Get all attestations for the valid time window of cutoffIndex up to index (as we just applied the pending attestations).
+	attestations := m.determineAttestationsFromWindow(cutoffIndex)
+	m.pendingAttestations.Evict(cutoffIndex)
+
+	// Store all attestations of cutoffIndex in bucketed storage via ads.Map / sparse merkle tree -> committed attestations.
+	tree, err := m.adsMapStorage(cutoffIndex)
+	if err != nil {
+		return 0, types.Identifier{}, errors.Wrapf(err, "failed to get attestation storage when committing slot %d", index)
+	}
+
+	// Add all attestations to the tree and calculate the new cumulative weight.
+	for _, a := range attestations {
+		// TODO: which weight are we using here? The current one? Or the one of the slot of the attestation/commitmentID?
```

Comment:
We need to come clear what committee we are using here, as you are requesting it "on demand": this should be the *TOTAL* committee always; in order words you should be always able to find the issuer's weight in the commitee; and since the weights are advanced only after a commitment, this should be fine. We should double-check that the sybilprotection vector is advanced only after any "weight-depending" committment activity has been performed!